contribute/design/proposal-cstor-orchestration.md

```diff
@@ -0,0 +1,437 @@
+       template:
+         metadata:
+           labels:
+             monitoring: volume_exporter_prometheus
+             openebs/controller: cstor-controller
+             pv: pvc-ee171da3-07d5-11e8-a5be-42010a8001be
+         spec:
+           containers:
+           - name: pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-ctrl
+             image: openebs/cstor:0.7.0               
+             ports:
+             - containerPort: 3260
+               protocol: TCP
+             resources: {}
+           - name: pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-ctrl-mgmt
+             image: openebs/m-cstor-mgmt:0.7.0
+             args:
+             - --volume-cr
+             - pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-volume
+             command:
+             - launch
+             ports:
+             - containerPort: 9500
+               protocol: TCP
+             resources: {}
+           - name: pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-ctrl-exporter
+             image: openebs/m-exporter:0.7.0               
+             ports:
+             - containerPort: 9500
+               protocol: TCP
+             resources: {}             
+       ```
+       The resources{} will be filled based on the resource (cpu, mem) requests and limits given in the Volume Policies associated with PVC. If nothing has been provided, Kubernetes will assign default values depending on the node resources. Please refer to the [Kubernetes Resource Limits and Request](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container).
+
+   * maya-apiserver will create the OV CStorVolumeReplicas as follows: 
+     * Query for the storage pools matching the given pool name in the PVC and pick up a subset of pools (based on the replica-count of the PVC). 
+     * For each replica, maya-apiserver will create - CStorVolumeReplica CR. This CStorVolumeReplica CR will contain:
+       - CStorPool (unique id like _7b99e406-1260-11e8-aa43-00505684eb2e_)
+       - Unique Name ( an hash will be suffixed to the PVC name like - _pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-rep-9440ab_
+       - Required capacity (obtained from PVC or StorageClass or default value)
+       - OV cStorService IP (obtained from cStorService _(pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-ctrl-service)_ CluserIP )
+       - status that can have values like pending, read-write, write-only, read-only depending on current state of replica. cstor-pool-mgmt will update the status with the correct value. 
+       
+       The YAML for the CStorVolumeReplica is as follows:
+       ```
+       apiVersion: openebs.io/v1alpha1
+       kind: CStorVolumeReplica
+       metadata:
+         name: pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-rep-9440ab
+       spec:
+         poolGUID: 7b99e406-1260-11e8-aa43-00505684eb2e
+         cstorControllerIP: <ip-address>
+         volumeName: demo-vol
```

Comment:
include volume id

Reply:
annotations:
    openebs.io/cstor-pool-guid: 7b99e406-1260-11e8-aa43-00505684eb2e
    openebs.io/cstor-vol-guid: 8b99e206-1260-11w8-aa43-005056d4eb23
like this? Also, this cstor-vol-guid will be obtained from zfs get guid cmd from sidecar and populated by sidecar along with status, only if vol creation is successful.

Reply:
i updated the uid fields to reflect the way, it is present on all other kubernetes objects like in the case of pod. 

```
       apiVersion: openebs.io/v1alpha1
       kind: CStorVolumeReplica
       metadata:
         name: pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-rep-9440ab
         uid: ee171da3-07d5-11e8-a5be-42010a8001be
         labels:
           "pool.openebs.io/uid" : 7b99e406-1260-11e8-aa43-00505684eb2e
           "pool.openebs.io/name" : pool1-84eb2e
       spec:
         cstorControllerIP: <ip-address>
         volumeName: demo-vol
         capacity: 5G
       status:
         phase: init
```

Reply:
@kmova Thanks for the update. will use labels instead of annotations. Also, will use pool.openebs.io/uid for zfs create and ignore pool.openebs.io/name and volumeName at sidecar level.

Reply:
yes

Reply:
Need to remove volumeName under spec and add as "volume.openebs.io/pvc-name" : demo-vol under labels, as discussed.